src/app/core/device.ts

```diff
@@ -0,0 +1,1036 @@
+        let cell = new Layer({ z_offset: 0, flip: false }, "cell");
+
+        this.addLayer(flowlayer);
+        this.addLayer(controllayer);
+
+        //TODO:Remove Cell layer from the whole system
+        this.addLayer(cell);
+
+        return [flowlayer, controllayer, cell];
+    }
+
+    /**
+     * Deletes the layer defined by the index
+     * @param {number} index
+     * @memberof Device
+     * @returns {void}
+     */
+    deleteLayer(index: number): void {
+        if (index != -1) {
+            this.__layers.splice(index, 1);
+        }
+    }
+
+    /**
+     * Returns the component identified by the id
+     * @param {string} id ID of the feature to get the component
+     * @return {Component|null} 
+     * @memberof Device
+     */
+    getComponentForFeatureID(id: string): Component | null {
+        for (let i in this.__components) {
+            let component = this.__components[i];
+            //go through each component's features
+            for (let j in component.features) {
+                let feature = component.features[j];
+                if (feature === id) {
+                    return component;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Generates a new new name for the type, use this to autogenerate the names for components that are typespecific
+     * @param {String} type
+     * @return {string}
+     * @memberof Device
+     */
+    generateNewName(type: string | String): string {
+        if (this.__nameMap.has(type)) {
+            let value = this.__nameMap.get(type);
+            this.__nameMap.set(type, value + 1);
+            return type + "_" + String(value + 1);
+        } else {
+            this.__nameMap.set(type, 1);
+            return type + "_1";
+        }
+    }
+
+    /**
+     * Returns a connection object corresponding to the ID
+     * @param {String} id ID of feature to get the connection
+     * @return {Connection|null} 
+     * @memberof Device
+     */
+    getConnectionForFeatureID(id: string | String): Connection | null {
+        for (let i in this.__connections) {
+            let connection = this.__connections[i];
+            //go through each component's features
+            for (let j in connection.features) {
+                let feature = connection.features[j];
+                if (feature.ID === id) {
+                    return connection;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Insert a connection between a valve component and the connection component
+     * @param {Valve} valve Valve object
+     * @param {Connection} connection Connection object
+     * @memberof Device
+     * @returns {void}
+     */
+    insertValve(valve: Valve, connection: Connection, is3D: boolean = false): void {
+        this.__valveMap.set(valve.getID(), connection.getID());
+        this.__valveIs3DMap.set(valve.getID(), is3D);
```

Comment:
Valve has no member function getID, nor does it appear to have a property ID.

Reply:
valve is of the type `Component`